src/api/index.md

```diff
@@ -0,0 +1,277 @@
+# Gutenberg API Prototype
+
+This is proposal for a block API for the new WordPress editor!
+
+Joen Asmussen pointed out that the design of the new editor would be
+"block first". This sunk in and spurred this adventure into creating an API that
+itself is just a single self referencing "block" that can create other blocks
+out of itself.
+
+## Overview
+
+### How does this thing work
+
+While trying to create a block that could make other blocks, which could make
+other blocks, with a nice syntax, I settled on the pattern seen in blockType.js
+et al. Basically, you have a function, that takes a function of properties and
+returns a new reference to a function assigned those properties. This is then
+partially applied to return a new function that takes properties. Each time the
+function is called a new reference is created to enable the ability for new
+entities to be created from a primitive version of itself carrying along its own
+properties and at the beginning starting with none.
+
+### Quick intro
+
+`blockType()`, `block()`, `control()`, and `controlType()`, all serve as an independent
+tool for building larger pieces. Each are identical other than their name. They
+all work from the same fundamental pattern. Let's look at `blockType`, and
+`controlType`. These would serve as a way to create factories for `block()` and
+control(). A good idea might be to have a `create()` method on a type, which would
+create the actual block, or control. Then each block or control would have a
+`render()` method which would create an actual representation of that block or
+control. `render()` could be used to return React components, DOM, vDOM, HTML,
+TinyMCE, whatever is chosen. Let's look at creating a `blockType`.
+
+### Creating a textBlockType
+
+```js
+// textBlock would be the block
+import textBlock from '../blocks/textBlock'
+import blockType from './blockType'
+
+const textBlockTypeProperties = {
+	type: 'wp-text',
+	title: 'Text',
+	controls: [
+		[
+			controlType( {
```

Comment:
What advantage is there to defining controls using the `controlType` wrapper function? In your [previous pull request reflections](https://github.com/WordPress/gutenberg/pull/48#issuecomment-279171574) you'd mentioned how this might help with composing controls, though I'm curious what the practical applications would be, and whether it's a common and painful enough problem that warrants additional APIs.

The alternative I see is simply passing each control as an object. And as far as composition goes, you could still extend one control with another through e.g. `Object.assign`.